File: writesave.php
<?php
/* Copyright (C) 2025 Joel Yliluoma - https://iki.fi/bisqwit */
/* License: BSD */

function Encode($d)
{
  $dict = is_object($d);
  if($dict)
  {
    // Change stdClass into Array
    $d = json_decode(json_encode($d), true, 512, JSON_OBJECT_AS_ARRAY);
  }
  elseif(is_array($d))
  {
    foreach($d as $k=>$v)
      if($k != (string)(int)$k)
      {
        $dict = true;
        break;
      }
  }
  if($dict)
  {
    // Dict
    if(count($d) <= 15)
      $r = chr(0x80 + count($d));
    else
      $r = chr(0xDE) . chr(count($d) >> 8) . chr(count($d) & 0xFF);
    foreach($d as $k=>$v)
    {
      $r .= Encode($k);
      $r .= Encode($v);
    }
    return $r;
  }
  if(is_string($d))
  {
    if(preg_match('@^byte (?:0x)?([0-9A-F]*)$@', $d, $mat))
      return chr(hexdec($mat[1]));
    if(strlen($d) >= 0x20)
      print("ERROR: Don't know how to encode strings longer than 31 characters\n");
    return chr(0xA0 + strlen($d)) . $d;
  }
  if(is_array($d))
  {
    if(count($d) <= 15)
      $r = chr(0x90 + count($d));
    else
      $r = chr(0xDC) . chr(count($d) >> 8) . chr(count($d) & 0xFF);
    foreach($d as $v)
      $r .= Encode($v);
    return $r;
  }
  if(is_float($d))
  {
    return chr(0xCA) . pack('G', $d);
  }
  if(is_int($d))
  {
    if($d >= 0 && $d <= 0x7F)
      return chr($d);
    elseif($d >= 0 && $d <= 0xFF)
      return chr(0xCC) . chr($d);
    elseif($d >= 0 && $d <= 0xFFFF)
      return chr(0xCD) . chr($d >> 8) . chr($d & 0xFF);
    else
    {
      print("ERROR: For integers, I only know the encoding for 0..65535. Don't know how $d works.\n");
      // Resort to encoding as a float
      return chr(0xCA) . pack('G', $d);
    }
  }
  print("ERROR: Unknown data type in {$d}\n");
}

if($argc < 2)
{
  $argv[] = '-h';
}
$outfn = '';
$force_overwrite = 0;
$accept_flags    = true;
$fake_write      = false;
foreach($argv as $a)
{
  if($a[0] == '-' && $accept_flags)
  {
    if($a == '-') $outfn = 'php://stdout';
    elseif($a == '--') { $accept_flags = false; }
    else for($n=1; $n<strlen($a); ++$n)
      switch($a[$n])
      {
        case 'f': $force_overwrite = 1; break;
        case 'i': $force_overwrite = -1; break;
        case 'n': $fake_write = true; break;
        default: print("Unknown option. ");
        case 'h':
          print("Usage: php writesave.php [-ifn] savefilename.sav < savedata.json\n");
          print("  -i Prompt before overwriting\n");
          print("  -f Always 'yes'\n");
          print("  -n Dry-run: Do not write any files\n");
          exit;
      }
  }
  else
    $outfn = $a;
}

$r = file_get_contents('php://stdin');
/* Strip comments */
$r = preg_replace('@/[*].*[*]/@m', '', $r);
/* Remove any occurrences of comma before ] or } */
$r = preg_replace('@,+\s*([\]\}])@', '\1', $r);

$data = json_decode($r);
$s = encode($data);
if($fake_write)
{
  print("Would write to $outfn:\n");
  $p = popen('hexdump -C', 'w');
  fwrite($p, $s);
  pclose($p);
}
else
{
  if(file_exists($outfn))
    switch($force_overwrite)
    {
      case -1:
      case 0:
        print("Output file already exists, cancelling\n");
        exit;
      case 1:
        break;
    }
  file_put_contents($outfn, $s);
}

File: writesave.py
# Copyright (C) 2025 Joel Yliluoma - https://iki.fi/bisqwit
# License: BSD
import sys
import re
import json
import struct
import os
import binascii

def encode(data):
    if isinstance(data, dict):
        n = len(data)
        if n <= 15:
            r = bytes([0x80 + n])
        else:
            r = bytes([0xDE, (n >> 8) & 0xFF, n & 0xFF])
        for k, v in data.items():
            r += encode(k)
            r += encode(v)
        return r
    elif isinstance(data, list):
        n = len(data)
        if n <= 15:
            r = bytes([0x90 + n])
        else:
            r = bytes([0xDC, (n >> 8) & 0xFF, n & 0xFF])
        for v in data:
            r += encode(v)
        return r
    elif isinstance(data, str):
        match = re.match(r'^byte\s*(?:0x)?([0-9A-Fa-f]+)$', data, re.IGNORECASE)
        if match:
            hex_val = match.group(1)
            try:
                byte_val = int(hex_val, 16)
                return bytes([byte_val])
            except ValueError:
                print("ERROR: Invalid hex in byte string\n")
                return b''
        utf8_data = data.encode('utf-8')
        if len(utf8_data) >= 32:
            print("ERROR: Don't know how to encode strings longer than 31 characters\n")
            return b''
        return bytes([0xA0 + len(utf8_data)]) + utf8_data
    elif isinstance(data, float):
        return bytes([0xCA]) + struct.pack('!f', data)
    elif isinstance(data, int):
        if 0 <= data <= 0x7F:
            return bytes([data])
        elif 0 <= data <= 0xFF:
            return bytes([0xCC, data])
        elif 0 <= data <= 0xFFFF:
            return bytes([0xCD, (data >> 8) & 0xFF, data & 0xFF])
        else:
            print(f"ERROR: For integers, I only know the encoding for 0..65535. Don't know how {data} works.\n")
            return bytes([0xCA]) + struct.pack('!f', data)
    else:
        print(f"ERROR: Unknown data type in {data}\n")
        return b''

def main():
    data = sys.stdin.read()
    
    data = re.sub(r'/\*.*?\*/', '', data, flags=re.DOTALL)
    data = re.sub(r',+\s*([\]\}])', r'\1', data)
    
    try:
        data = json.loads(data)
    except json.JSONDecodeError as e:
        print(f"JSON decode error: {e}")
        sys.exit(1)
    
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('-i', action='store_true', help='Prompt before overwriting')
    parser.add_argument('-f', action='store_true', help='Always overwrite')
    parser.add_argument('-n', action='store_true', help='Dry-run')
    parser.add_argument('output', nargs='?', help='Output filename')
    
    args = parser.parse_args()
    
    force_overwrite = -1 if args.i else (1 if args.f else 0)
    outfn = args.output
    
    if outfn == '-':
        outfn = 'php://stdout'
    
    s = encode(data)
    
    if args.n:
        print(f"Would write to {outfn}:")
        print(binascii.hexlify(s).decode('ascii'))
    else:
        if outfn == 'php://stdout':
            sys.stdout.buffer.write(s)
        else:
            if os.path.exists(outfn) and force_overwrite != 1:
                if force_overwrite == -1:
                    pass
                print("Output file already exists, cancelling")
                sys.exit(1)
            with open(outfn, 'wb') as f:
                f.write(s)

if __name__ == '__main__':
    main()
